test/test_helper.rb

```diff
@@ -5,72 +5,302 @@
-    )
-end
+                @installation_manifest =
+                    Autoproj::InstallationManifest.new(@installation_manifest_path)
+
+                installation_manifest.save
+                return ws unless vcs
+
+                ws.manifest.vcs = Autoproj::VCSDefinition.from_raw(vcs)
+                ws.config.set 'manifest_source', vcs.dup, true
+                ws.config.save
+
+                autoproj_daemon_git_init('autoproj', dummy: false)
+                ws
+            end
+
+            def autoproj_daemon_mock_github_api
+                @mock_client = flexmock(Octokit::Client).new_instances
+                @mock_rate_limit = flexmock
+
+                @mock_client.should_receive(:rate_limit).and_return { @mock_rate_limit }
+                @mock_client.should_receive(:rate_limit!).and_return { @mock_rate_limit }
+                @mock_rate_limit
+                    .should_receive(:remaining)
+                    .and_return { @client_rate_limit_remaining }
+
+                @mock_client
+                    .should_receive(:pull_requests)
+                    .and_return do |repo, _options|
+                        @client_pull_requests[repo] || []
+                    end
+
+                @mock_client
+                    .should_receive(:pull_request)
+                    .and_return do |repo, number, _options|
+                        (@client_pull_requests[repo] || [])
+                            .find { |pr| pr['number'] == number }
+                    end
+
+                @mock_client
+                    .should_receive(:branches)
+                    .and_return do |repo, _options|
+                        @client_branches[repo] || []
+                    end
+
+                @mock_client
+                    .should_receive(:user)
+                    .and_return do |user|
+                        @client_users[user] || {}
+                    end
+
+                @mock_client
+                    .should_receive(:user_events)
+                    .and_return do |user|
+                        @client_user_events[user] || []
+                    end
+
+                @mock_client
+                    .should_receive(:organization_events)
+                    .and_return do |organization|
+                        @client_organization_events[organization] || []
+                    end
+
+                @client_rate_limit_remaining = 1000
+                @client_pull_requests = {}
+                @client_branches = {}
+                @client_users = {}
+                @client_organization_events = {}
+                @client_user_events = {}
+            end
+
+            def autoproj_daemon_git_init(dir, dummy: true)
+                dir = File.join(@ws.root_dir, dir)
+                FileUtils.mkdir_p dir if dummy
+
+                Dir.chdir(dir) do
+                    FileUtils.touch(File.join(dir, 'dummy')) if dummy
+
+                    system('git init > /dev/null')
```

Comment:
The more robust way to do this (because no need for escaping, no need to `chdir` and no need for a shell) would be

~~~
system('git', 'init', out: '/dev/null', chdir: dir)
~~~

add `err: '/dev/null'` to also redirect stderr.

Moreover, you definitely should be checking the return value of `system`

Reply:
Fixed.

Used popen3 instead because I think it's better if one wants to handle the output. Kept `Dir.chdir` though because I need `FileUtils.touch` in there as well

Reply:
Dir.chdir's fine here, because tests and all, but FYI it's fragile (concurrent chdir's will interfere with each other)

open3 accepts all arguments that `spawn` does, so `chdir` as well. The touch is using a full path, so it does not need the chdir either.

But that's fine here... Do as you please.

Reply:
> The touch is using a full path, so it does not need the chdir either.

Completely missed that somehow.. Fixed. Thanks! :+1: